### PR TITLE
Replace `#ifdef __CUDA_ARCH__` in `cast_{to,from}_half()`

### PR DESCRIPTION
**Pull request by dalg24**
I only dealt with the CUDA half conversions.  My plan was to deal with
https://github.com/kokkos/kokkos/blob/5c22f4b41c41d52be165f777db5f5a3e5fb7c4d5/core/src/Kokkos_Half.hpp#L55-L61
later

### CODE REVIEW

**Review by e10harvey — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

File: core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp [hunk 1]
```diff
@@ -163,62 +145,52 @@ cast_from_half(half_t val) {
...
 KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, int>::value, T>
 cast_from_half(half_t val) {
-#ifdef __CUDA_ARCH__
-  return __half2int_rz(half_t::impl_type(val));
-#else
-  return static_cast<T>(__half2float(half_t::impl_type(val)));
-#endif
+  KOKKOS_IF_ON_DEVICE((
+
+      return __half2int_rz(half_t::impl_type(val));))
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp:166]: I kinda messed up the format here...

Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp:166]:
```
git diff f8ef5fd 9e081e9
```
yields
```diff
diff --git a/core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp b/core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp
index 274322630..40a263561 100644
--- a/core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp
+++ b/core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp
@@ -161,9 +161,7 @@ KOKKOS_INLINE_FUNCTION
 template <class T>
 KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, int>::value, T>
 cast_from_half(half_t val) {
-  KOKKOS_IF_ON_DEVICE((
-
-      return __half2int_rz(half_t::impl_type(val));))
+  KOKKOS_IF_ON_DEVICE((return __half2int_rz(half_t::impl_type(val));))
   KOKKOS_IF_ON_HOST(
       (return static_cast<T>(__half2float(half_t::impl_type(val)));))
 }
```
Failure was SYCL perf test

**Comment by dalg24** [comment]
Retest this please

**Comment by masterleinad** [comment]
It's again only the `SYCL` PerfTest failing... I will have a look if there is anything that changed.

**Pull request merged**